server/sonar-server/src/main/java/org/sonar/server/computation/step/LoadCrossProjectDuplicationsRepositoryStep.java

```diff
@@ -0,0 +1,169 @@
+  public String getDescription() {
+    return "Feed duplications repository with cross project duplications";
+  }
+
+  private class CrossProjectDuplicationVisitor extends TypeAwareVisitorAdapter {
+
+    private CrossProjectDuplicationVisitor() {
+      super(CrawlerDepthLimit.FILE, PRE_ORDER);
+    }
+
+    @Override
+    public void visitFile(Component file) {
+      visitComponent(file);
+    }
+
+    private void visitComponent(Component component) {
+      List<CpdTextBlock> cpdTextBlocks = newArrayList(reportReader.readCpdTextBlocks(component.getReportAttributes().getRef()));
+      LOGGER.trace("Found {} cpd blocks on file {}", cpdTextBlocks.size(), component.getKey());
+      if (cpdTextBlocks.isEmpty()) {
+        return;
+      }
+
+      Collection<String> hashes = from(cpdTextBlocks).transform(CpdTextBlockToHash.INSTANCE).toList();
+      List<DuplicationUnitDto> dtos = selectDuplicates(component, hashes);
+      Collection<Block> duplicatedBlocks = from(dtos).transform(DtoToBlock.INSTANCE).toList();
+      Collection<Block> originBlocks = from(cpdTextBlocks).transform(new CpdTextBlockToBlock(component.getKey())).toList();
+
+      integrateCrossProjectDuplications.computeCpd(component, originBlocks, duplicatedBlocks);
+    }
+
+    private List<DuplicationUnitDto> selectDuplicates(Component file, Collection<String> hashes) {
+      DbSession dbSession = dbClient.openSession(false);
+      try {
+        Snapshot projectSnapshot = analysisMetadataHolder.getBaseProjectSnapshot();
+        Long projectSnapshotId = projectSnapshot == null ? null : projectSnapshot.id();
+        return dbClient.duplicationDao().selectCandidates(dbSession, projectSnapshotId, file.getFileAttributes().getLanguageKey(), hashes);
+      } finally {
+        dbClient.closeSession(dbSession);
+      }
+    }
+  }
+
+  private enum CpdTextBlockToHash implements Function<CpdTextBlock, String> {
+    INSTANCE;
+
+    @Override
+    public String apply(@Nonnull CpdTextBlock duplicationBlock) {
+      return duplicationBlock.getHash();
+    }
+  }
+
+  private enum DtoToBlock implements Function<DuplicationUnitDto, Block> {
+    INSTANCE;
+
+    @Override
+    public Block apply(@Nonnull DuplicationUnitDto dto) {
+      return Block.builder()
+        .setResourceId(dto.getComponentKey())
+        .setBlockHash(new ByteArray(dto.getHash()))
+        .setIndexInFile(dto.getIndexInFile())
+        .setLines(dto.getStartLine(), dto.getEndLine())
+        .build();
+    }
+  }
+
+  private static class CpdTextBlockToBlock implements Function<CpdTextBlock, Block> {
+    private final String fileKey;
+    private int indexInFile = 0;
+
+    public CpdTextBlockToBlock(String fileKey) {
+      this.fileKey = fileKey;
+    }
+
+    @Override
+    public Block apply(@Nonnull CpdTextBlock duplicationBlock) {
+      return Block.builder()
+        .setResourceId(fileKey)
+        .setBlockHash(new ByteArray(duplicationBlock.getHash()))
+        .setIndexInFile(indexInFile++)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Extract this increment or decrement operator into a dedicated statement. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS881)